dwave/preprocessing/src/presolveimpl.hpp

```diff
@@ -928,6 +851,123 @@ class PresolverImpl {
         std::vector<Transform> transforms_;
     };
 
+    template <dimod::Sense Sense>
+    bool technique_domain_propagation(const constraint_type& constraint) {
+        // Dev note: if-branche(s) below rely on this static assert
+        static_assert(Sense == dimod::Sense::GE || Sense == dimod::Sense::LE,
+                      "Sense must be <= or >=; equality constraints should call both");
+
+        assert(constraint.sense() == dimod::Sense::EQ || constraint.sense() == Sense);
+
+        // todo: extend to quadratic
+        if (!constraint.is_linear()) {
+            return false;  // no changes
+        }
+
+        // Cannot use soft constraints to strengthen bounds.
+        if (constraint.is_soft()) {
+            return false;  // no changes
+        }
+
+        // In order to not fall into zeno's paradox here, we don't shrink bounds when
+        // the reduction is small.
+        static constexpr double BOUND_CHANGE_MINIMUM = 1.0e-3 * FEASIBILITY_TOLERANCE;
+
+        // Very large activities will create numeric issues. So we don't do domain
+        // propagation on those
+        static constexpr double MAX_ACTIVITY = 1.0e10;
+
+        // In the following code there are three cases we want to account for:
+        // - No large activities: we can do domain propagation on all of the variables
+        // - Exactly one large activity: we can do domain propagation on the one variable with
+        //   large activity
+        // - 2+ large activities: we can't do domain propagation
+
+        // Precalculate the activities of each of the variables
+        std::vector<bias_type> activities;
+        std::vector<index_type> large_activities;
+        for (const auto& v : constraint.variables()) {
+            if constexpr (Sense == dimod::Sense::LE) {
+                activities.emplace_back(minimal_activity(constraint, v));
+            } else {  // Sense == dimod::Sense::GE, enforced by static_assert above
+                activities.emplace_back(maximal_activity(constraint, v));
+            }
+
+            if (std::abs(activities.back()) > MAX_ACTIVITY) {
+                large_activities.emplace_back(v);
+
+                // Early exit test: in the case that we have more than one large
+                // activity, we cannot do domain propagation at all
+                // Everything will work without this test, but it can save some time
+                if (large_activities.size() > 1) {
+                    return false;  // no changes
+                }
+
+                // Since we've marked it as large, let's zero it out to make
+                // some of the sums later a bit safer
+                activities.back() = 0;
+            }
+        }
+
+        // Get the total activity of everything (except the large activity variables)
+        const bias_type activity =
```

Comment:
Could use `std::accumulate()` for more readability, or calculate activities using a prefix sum in the previous loop as we visit each variable. But it shouldn't make a huge difference anyway.